core/modules/ip_checksum.cc

```diff
@@ -37,12 +37,28 @@
 void IPChecksum::ProcessBatch(bess::PacketBatch *batch) {
   using bess::utils::Ethernet;
   using bess::utils::Ipv4;
+  using bess::utils::Vlan;
+  using bess::utils::be16_t;
 
   int cnt = batch->cnt();
 
   for (int i = 0; i < cnt; i++) {
     Ethernet *eth = batch->pkts()[i]->head_data<Ethernet *>();
-    Ipv4 *ip = reinterpret_cast<Ipv4 *>(eth + 1);
+    Ipv4 *ip;
+
+    // Calculate checksum only for IPv4 packets
+    if (eth->ether_type == be16_t(Ethernet::Type::kIpv4)) {
+      ip = reinterpret_cast<Ipv4 *>(eth + 1);
+    } else if (eth->ether_type == be16_t(Ethernet::Type::kVlan)) {
```

Comment:
Could you check for double tagged packets, too?

Reply:
Ok, done, thanks!